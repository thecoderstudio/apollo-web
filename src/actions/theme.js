function changeTheme(theme) {
  return {
    type: 'CHANGE_THEME',
    theme
  };
};

export { changeTheme };
