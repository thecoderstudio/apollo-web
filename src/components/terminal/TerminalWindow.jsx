import React from 'react';
import styled, { withTheme } from 'styled-components';
import { Rnd } from 'react-rnd';
import Card from '../../components/Card';
import Terminal, { openTerminal } from './Terminal';

const Window = styled(Card)`
  padding: 0px;
  border: ${props => `1px solid ${props.statusColor}`};
  border-style: none none solid none;
`;

const TaskBar = styled.div`
  display: flex;
  height: 30px;
  align-items: center;
`;

const Controls = styled.div`
  flex: 1;
  display: flex;
  justify-content: flex-end;
`;

const WindowButton = styled.div`
  height: 12px;
  width: 12px;
  margin-right: 15px;
  background-color: ${props => props.color}; 
  border-radius: 50%;
  cursor: pointer;
`;

const Space = styled.div`
  flex: 1;
`;

const Title = styled.h4`
  flex: 1;
  width: 100%;
  text-align: center;
`;

class TerminalWindow extends React.PureComponent {
  constructor(props) {
    super(props);
    this.openTerminalInNewWindow = this.openTerminalInNewWindow.bind(this);
    this.close = this.close.bind(this);
    this.terminalRef = React.createRef();
  }

  componentDidMount() {
    this.terminalRef.current.fit();
  }

  close() {
    this.props.onClose();
  }

  openTerminalInNewWindow() {
    openTerminal(this.props.agent.id);
  }

  render() {
    let statusColor = this.props.theme.red;
    if (this.props.agent.connection_state == 'connected') {
      statusColor = this.props.theme.green;
    }

    return (
      <Rnd
        default={{
          x: 0,
          y: 0,
          width: 700,
          height: 400
        }}
        minWidth={700}
        minHeight={300}
        bounds="window"
        onResize={(e, direction, ref, delta, position) => {
          this.terminalRef.current.fit();
        }}
      >
        <Window statusColor={statusColor}>
          <TaskBar>
            <Space />
            <Title>{this.props.agent.name}</Title>
            <Controls>
              <WindowButton id="expand-button" onClick={this.openTerminalInNewWindow} color={this.props.theme.green} />
              <WindowButton id="close-button" onClick={this.close} color={this.props.theme.red} />
            </Controls>
          </TaskBar>
          <Terminal agent={this.props.agent} ref={this.terminalRef} />
        </Window>
      </Rnd>
    );
  }
}

export default withTheme(TerminalWindow);
