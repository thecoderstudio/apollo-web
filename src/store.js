import { createStore } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'
import storage from 'redux-persist/lib/storage';
import mapTransformer from './transformers/map';
import rootReducer from './reducers/root';

const persistConfig = {
  key: 'root',
  whitelist: ['auth', 'agent'],
  transforms: [mapTransformer({ whitelist: 'agent' })],
  storage
};

const persistedReducer = persistReducer(persistConfig, rootReducer);
export const store = createStore(persistedReducer);
export const persistor = persistStore(store);
