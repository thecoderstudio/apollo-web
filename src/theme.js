export const darkTheme = {
  name: "dark",
  black: "#181818",
  lightBlack: "#1B1B1B",
  white: "#F8F8F8",
  primary: "#02A3EE",
  accent: "#4D02EE",
  connectedColor: "#27AE60",
  disconnectedColor: "#C0392B",
  connectingColor: "#F39C12",
  error: "#ED3E2A"
};

function getBrowserPreferredTheme() {
  return darkTheme;
}

export const browserPreferredTheme = getBrowserPreferredTheme();
