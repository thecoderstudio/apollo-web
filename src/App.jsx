import React from 'react';
import { connect } from 'react-redux';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import styled, { createGlobalStyle, ThemeProvider } from 'styled-components';
import ProtectedRoute from './components/ProtectedRoute';
import NavBar from './components/NavBar';
import Notifications from './components/notifications/Notifications';
import Dashboard from './pages/Dashboard';
import Welcome from './pages/Welcome';
import AgentDetail from './pages/AgentDetail';
import TerminalPage from './pages/TerminalPage';
import Admin from './pages/Admin';
import NotFound from './pages/NotFound';
import { darkTheme } from './theme';

const GlobalStyle = createGlobalStyle`
  @import url('https://fonts.googleapis.com/css2?family=B612:wght@400;700&family=Roboto:wght@300;400&display=swap');

  html, body {
    margin: 0px;
    padding: 0px;
    height: 100%;
    width: 100%;
    font-weight: 500;
    color: ${(props) => props.theme.white};
    background-color: ${(props) => props.theme.black};
  }

  body {
    font-family: 'Roboto', sans-serif;
  }

  h1, h2, h3, h4, h5, h6 {
    font-family: 'B612', sans-serif;
  }
`;

const Content = styled.div`
  height: 100%;
  width: 100%;
`;

function App(props) {
  return (
    <div>
      <ThemeProvider theme={darkTheme}>
        <BrowserRouter>
          <Content>
            {props.authenticated && props.hasChangedPassword && <NavBar />}
            <Switch>
              <ProtectedRoute exact path="/" component={Dashboard} fallbackComponent={Welcome} />
              <ProtectedRoute exact path="/agent/:agentId/shell" component={TerminalPage} fallbackComponent={NotFound} />
              <ProtectedRoute exact path="/agent/:agentId" component={AgentDetail} fallbackComponent={NotFound} />
              <ProtectedRoute exact path="/admin" component={Admin} fallbackComponent={NotFound} role="admin" />
              <Route component={NotFound} />
            </Switch>
            <Notifications />
            <GlobalStyle />
          </Content>
        </BrowserRouter>
      </ThemeProvider>
    </div>
  );
}

export default connect(
  (state) => ({
    authenticated: state.authenticated,
    hasChangedPassword: state.currentUser.hasChangedInitialPassword
  })
)(App);
