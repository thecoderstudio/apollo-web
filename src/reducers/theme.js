import { browserPreferredTheme } from '../theme';

export default function themeReducer(state = { theme: browserPreferredTheme }, action) {
  switch (action.type) {
    case 'CHANGE_THEME':
      return { theme: action.theme };
    default:
      return state;
  }
}
