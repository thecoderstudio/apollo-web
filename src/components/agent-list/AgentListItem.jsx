import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../Text';
import ConnectionState from './ConnectionState';
import { openTerminal } from '../terminal/Terminal';
import TerminalWindow from '../terminal/TerminalWindow';
import MobileChecker from '../../util/MobileChecker';
import media from '../../util/media';

const propTypes = {
  agent: PropTypes.object.isRequired
};

const Container = styled.li`
  display: flex;
  justify-content: flex-end;
  align-items: center;

  border-radius: 8px;
	border: 1px solid white;

  height: 30px;
	line-height: 30px;
	padding: 15px 30px;
  margin-top: 25px;

	${
    media.phone`
      flex-direction: column;
      height: auto;
    `
  }
`;

const StyledText = styled(Text)`
  width: 100%;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  
    ${
      media.phone`
        text-align: center;
        width: 100%;
      `
    }
`;

const Controls = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 250px;
  

	${
    media.phone`
      width: 100%;
    `
  }
`;

const Icon = styled.i`
  cursor: ${props => props.active ? 'pointer' : 'inherit'};
`;

const TerminalIcon = styled(Icon)`
  color: ${props => props.active ? props.theme.white : props.theme.inactive};
`;

export default class AgentListItem extends React.PureComponent {
  constructor(props) {
    super(props);
    this.openTerminal = this.openTerminal.bind(this);
    this.createTerminal = this.createTerminal.bind(this);
    this.closeTerminal = this.closeTerminal.bind(this);
    this.state = {terminalOpen: false};
  }

  openTerminal() {
    if (new MobileChecker().isMobile ){
      openTerminal(this.props.agent.id);
      return;
    }
    this.setState({terminalOpen: true});
  }

  createTerminal() {
      return <TerminalWindow agent={this.props.agent} onClose={this.closeTerminal} />;
  }

  closeTerminal() {
    this.setState({terminalOpen: false});
  }

  render() {
    const connected = this.props.agent.connection_state == 'connected';
    let terminal;
    if (this.state.terminalOpen) {
      terminal = this.createTerminal();
    }

    return (
      <Container>
        <StyledText>{this.props.agent.name}</StyledText>
          <Controls>
            <ConnectionState connectionState={this.props.agent.connection_state} />
            <TerminalIcon active={connected} onClick={this.openTerminal} className="fas fa-terminal" />
          </Controls>
          {terminal}
      </Container>
    );
  }
}

AgentListItem.propTypes = propTypes;
