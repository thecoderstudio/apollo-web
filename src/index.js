import React from 'react';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { render } from 'react-dom';
import App from './App';
import { store, persistor } from './store';

render(
  <Provider store={store}>
    <PersistGate loading={null} persistor={persistor}>
      <App />
    </PersistGate>
  </Provider>,
  document.getElementById('app')
);
