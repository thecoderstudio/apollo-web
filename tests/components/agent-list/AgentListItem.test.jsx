import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import WS from 'jest-websocket-mock';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AgentListItem from '../../../src/components/agent-list/AgentListItem';
import TerminalWindow from '../../../src/components/terminal/TerminalWindow';
import { darkTheme } from '../../../src/theme';
import MobileChecker from '../../../src/util/MobileChecker';

Enzyme.configure({ adapter: new Adapter() });

function getComponentTags(connection_state) {
  return (
    <div>
      <ThemeProvider theme={darkTheme}>
        <AgentListItem agent={{
          id: "fakeid",
          name: "agentName",
          connection_state: connection_state
        }} />
      </ThemeProvider>
    </div>
  );
}

function getComponent(connection_state) {
  return renderer.create(getComponentTags(connection_state));
}

describe('agent list item', () => {
  const server = new WS(`ws://localhost:1234/agent/a2346886-83ba-442d-9fb7-d024c6274e22/shell`);
  var wrapper;

  beforeEach(() => {
    process.env = {
      APOLLO_WS_URL: 'ws://localhost:1234/'
    };
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
    wrapper = mount(getComponentTags("connected"));
  });

  afterEach(() => {
    WS.clean();
  });

  it('renders connected correctly', () => {
    const tree = getComponent("connected").toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders disconnected correctly', () => {
    const tree = getComponent("disconnected").toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('opens and closes the terminal', () => {
    wrapper.find('i').simulate('click');
    expect(wrapper.find(TerminalWindow)).toHaveLength(1);

    // Close is supposed to call a callback given by the list item.
    wrapper.find('#close-button').at(0).simulate('click');
    expect(wrapper.find(TerminalWindow)).toHaveLength(0);
  });

  it('opens terminal in new window on mobile', () => {
    const location = window.location;
    const expectedHref = `${location.protocol}//${location.host}/agent/fakeid/shell`
    global.open = jest.fn();

    const mock = jest.fn();
    mock.mockReturnValueOnce(true);
    new MobileChecker().isMobile = mock;
    wrapper.find('i').simulate('click');

    expect(global.open).toHaveBeenCalledWith(expectedHref);
  });
});
