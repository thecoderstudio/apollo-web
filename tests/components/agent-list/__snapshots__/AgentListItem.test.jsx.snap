// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent list item renders connected correctly 1`] = `
<div>
  <li
    className="sc-fznyAO czxrNm"
  >
    <span
      className="sc-AxjAm sc-fznKkj eKuYEa"
    >
      agentName
    </span>
    <div
      className="sc-fznZeY gwCZdU"
    >
      <div
        className="sc-AxirZ hLbodF"
      >
        <div
          className="sc-AxiKw cDPPUu"
        />
        <span
          className="sc-AxjAm sc-AxhCb fnilXu"
        >
          Connected
        </span>
      </div>
      <i
        className="sc-fzokOt sc-fzqBZW gjFXho fas fa-terminal"
        onClick={[Function]}
      />
    </div>
  </li>
</div>
`;

exports[`agent list item renders disconnected correctly 1`] = `
<div>
  <li
    className="sc-fznyAO czxrNm"
  >
    <span
      className="sc-AxjAm sc-fznKkj eKuYEa"
    >
      agentName
    </span>
    <div
      className="sc-fznZeY gwCZdU"
    >
      <div
        className="sc-AxirZ hLbodF"
      >
        <div
          className="sc-AxiKw gYdmwA"
        />
        <span
          className="sc-AxjAm sc-AxhCb bSClzs"
        >
          Disconnected
        </span>
      </div>
      <i
        className="sc-fzokOt sc-fzqBZW hmiIRt fas fa-terminal"
        onClick={[Function]}
      />
    </div>
  </li>
</div>
`;
